### PR TITLE
[Distributed] Don't crash in thunk generation when missing SR conformance

### PR DESCRIPTION
**Pull request by ktoso**
**Problem:** We would properly diagnose the missing conformance; but the compiler would still try to emit a thunk for the errored `func`. Tests which only `-typecheck -verify` would not catch this issue because it is _after_ type checking.

**The fix:** This prevents distributed thunks from being generated when the func they are based on already has errors -- otherwise we can run into missing types and null pointers as we try to generate the thunk.
a
**Radar:** rdar://116261701

And found a follow up we should do: rdar://117677422

### CODE REVIEW

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckDistributed.cpp [hunk 1]
```diff
@@ -498,8 +498,22 @@ bool CheckDistributedFunctionRequest::evaluate(
...
-        C, extension->getGenericRequirements());
+    auto actorOrProtocol = extension->getExtendedNominal();
+    if (auto actor = dyn_cast<ClassDecl>(actorOrProtocol)) {
+      assert(actor->isAnyActor());
+      serializationRequirements = getDistributedSerializationRequirementProtocols(
+          getDistributedActorSystemType(actor)->getAnyNominal(),
+          C.getProtocol(KnownProtocolKind::DistributedActorSystem));
+    } else if (auto protocol = dyn_cast<ProtocolDecl>(actorOrProtocol)) {
+      extractDistributedSerializationRequirements(
+          C, protocol->getGenericRequirements(),
```
Comment (xedin) [inline lib/Sema/TypeCheckDistributed.cpp:509]: I don't think you need to do this because `extension->getGenericRequirements()` should cover both.

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test macOS

Failure was unrelated:

fatal error: module map file '/home/build-user/build/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux-gnu/release/llbuildBasic.build/module.modulemap' not found [ci.swift.org/job/swift-PR-macos-smoke-test/9867/console](https://ci.swift.org/job/swift-PR-macos-smoke-test/9867/console)

**Comment by ktoso** [comment]
@swift-ci please smoke test macOS

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
This is looking better now, thanks @slavapestov 

The only case we've regressed on now is:

```

extension NoSerializationRequirementYet
  where SerializationRequirement: Codable {
  // expected-error@+1{{result type 'NotCodable' of distributed instance method 'test4' does not conform to serialization requirement 'Codable'}}
  distributed func test4() -> NotCodable {
    .init()
  }
}
```

since the `getConcreteReplacementForMemberSerializationRequirement` will give a null type for this... I'm not sure how to handle this, can you advise?

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please clean smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: lib/AST/DistributedDecl.cpp [hunk 2]
```diff
@@ -106,17 +107,28 @@ Type swift::getConcreteReplacementForMemberSerializationRequirement(
...
       signature = DC->getGenericSignatureOfContext();
     }
 
-    auto SerReqAssocType = DA->getAssociatedType(C.Id_SerializationRequirement)
-                               ->getDeclaredInterfaceType();
+    // Also store all `SerializationRequirement : SomeProtocol` requirements
+    for (auto requirement: signature.getRequirements()) {
+      if (requirement.getFirstType()->isEqual(SerReqAssocType) &&
+          requirement.getKind() == RequirementKind::Conformance) {
+        if (auto nominal = requirement.getSecondType()->getAnyNominal()) {
```
Comment (slavapestov) [inline lib/AST/DistributedDecl.cpp:125]: The RHS of a conformance requirement is always a protocol. Call requirement.getProtocolDecl() to get it

Comment (ktoso) [reply lib/AST/DistributedDecl.cpp:125]: Nice, thank you!

Comment (ktoso) [reply lib/AST/DistributedDecl.cpp:125]: done in https://github.com/apple/swift/pull/69874

**Pull request merged**